vm/src/stdlib/syslog.rs

```diff
@@ -0,0 +1,128 @@
+
+    #[pyattr]
+    use libc::{
+        LOG_ALERT, LOG_AUTH, LOG_CONS, LOG_CRIT, LOG_DAEMON, LOG_DEBUG, LOG_EMERG, LOG_ERR,
+        LOG_INFO, LOG_KERN, LOG_LOCAL0, LOG_LOCAL1, LOG_LOCAL2, LOG_LOCAL3, LOG_LOCAL4, LOG_LOCAL5,
+        LOG_LOCAL6, LOG_LOCAL7, LOG_LPR, LOG_MAIL, LOG_NDELAY, LOG_NEWS, LOG_NOTICE, LOG_NOWAIT,
+        LOG_ODELAY, LOG_PID, LOG_SYSLOG, LOG_USER, LOG_UUCP, LOG_WARNING,
+    };
+
+    #[cfg(not(target_os = "redox"))]
+    #[pyattr]
+    use libc::{LOG_AUTHPRIV, LOG_CRON, LOG_PERROR};
+
+    fn get_argv(vm: &VirtualMachine) -> Option<PyStrRef> {
+        if let Some(argv) = vm.state.settings.argv.first() {
+            if !argv.is_empty() {
+                return Some(
+                    PyStr::from(match argv.find('\\') {
+                        Some(value) => &argv[value..],
+                        None => argv,
+                    })
+                    .into_ref(vm),
+                );
+            }
+        }
+        None
+    }
+
+    static SYSLOG_IDENT: AtomicCell<Option<Box<std::ffi::CStr>>> = AtomicCell::new(None);
+    static SYSLOG_IS_OPEN: AtomicCell<bool> = AtomicCell::new(false);
+
+    #[derive(Default, FromArgs)]
+    struct OpenLogArgs {
+        #[pyarg(any, optional)]
+        ident: OptionalOption<PyStrRef>,
+        #[pyarg(any, optional)]
+        logoption: OptionalArg<i32>,
+        #[pyarg(any, optional)]
+        facility: OptionalArg<i32>,
+    }
+
+    #[pyfunction]
+    fn openlog(args: OpenLogArgs, vm: &VirtualMachine) -> PyResult<()> {
+        let logoption = args.logoption.unwrap_or(0);
+        let facility = args.facility.unwrap_or(LOG_USER);
+        let ident = match args.ident.flatten() {
+            Some(args) => {
+                let cstr = args.to_cstring(vm)?.into_boxed_c_str();
+                let cstr_ptr = cstr.as_ptr();
+                SYSLOG_IDENT.store(Some(cstr));
+                cstr_ptr
+            }
+            None => match get_argv(vm) {
+                Some(argv) => {
+                    let cstr = argv.to_cstring(vm)?.into_boxed_c_str();
+                    let cstr_ptr = cstr.as_ptr();
+                    SYSLOG_IDENT.store(Some(cstr));
+                    cstr_ptr
+                }
+                None => std::ptr::null(),
+            },
+        };
+
+        unsafe { libc::openlog(ident, logoption, facility) };
+        SYSLOG_IS_OPEN.store(true);
+        Ok(())
+    }
+
+    #[derive(FromArgs)]
+    struct SysLogArgs {
+        #[pyarg(positional)]
+        priority: PyObjectRef,
+        #[pyarg(positional, optional)]
+        message_object: OptionalOption<PyStrRef>,
+    }
+
+    #[pyfunction]
+    fn syslog(args: SysLogArgs, vm: &VirtualMachine) -> PyResult<()> {
+        let (priority, msg) = match args.message_object.flatten() {
+            Some(s) => (i32::try_from_object(vm, args.priority)?, s),
+            None => (LOG_INFO, PyStrRef::try_from_object(vm, args.priority)?),
+        };
+
+        if !SYSLOG_IS_OPEN.load() {
```

Comment:
I am sorry. I messaged you it look like able to be managed with AtomicCell, but it doesn't. `closelog` can be called betwen line 95-96 and then SYSLOG_IDENT will be corrupted. Would you change it to RwLock?

Reply:
That's the point for lazy_static or OnceCell.

```rust
static SYSLOG_IDENT: OnceCell<PyRwLock<Option<Box<std::ffi::CStr>>>> = OnceCell::new();
```

https://docs.rs/once_cell/1.8.0/once_cell/#lazy-initialized-global-data

Creating a helper function will be convenient.

```rust
fn syslog_global_ident() -> &'static PyRwLock<Option<Box<CStr>>> {
    static SYSLOG_IDENT: OnceCell<PyRwLock<Option<Box<std::ffi::CStr>>>> = OnceCell::new();
    SYSLOG_INIT.get_ot_init(|| ...)
}
```

Then both open and close function can access to the shared global ident through the function like `syslog_global_ident().write()` or ``syslog_global_ident().read()`

Reply:
Oh my gosh. this makes me shock 😂. I applied your suggestion.

Reply:
uhm... There was no error in my environment. I'll solve CI error before sleeping 😢